### PR TITLE
mock and skip mechanism for act

### PR DESCRIPTION
**Pull request by Ginxo**
* publish.yml workflow covered
* some secondary functions/utilities have been moved to new act-helper.ts file
* skip and mock mechanisms for workflow steps added
* GithubWorkflowStep type representing github workflow yaml file added

### CODE REVIEW

File: src/act/act.ts [hunk 1]
```diff
@@ -45,36 +47,47 @@ export class Act {
...
     return this.run([event], opts);
   }
 
+  private async handleStepsBeforeRunningJobs(workflowFilter: WorkflowFilter, opts?: RunOpts) {
+    if (opts?.steps?.skip?.length || opts?.steps?.mock?.length) {
+      const jobs = await this.list({ workflowFilter });
+      if (!jobs?.length) {
+        throw new Error(`There are no jobs for filter ${workflowFilter}.`)
+      }
+      treatSteps(jobs, path.join(this.cwd, ".github", "workflows"), opts.steps)
```
Comment (Ginxo) [inline src/act/act.ts:87]: @shubhbapna is this ".github/workflows" path always like that?

File: test/.github/workflows/publish.test.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+            "PUBLISH_REPO_NAME": {
+                files: FileUtils.listMockGithubFiles(projectRootFolder, "repo")
+            },
+        },
+    });
+    await mockGithub.setup();
+    const act = new Act(mockGithub.repo.getPath("PUBLISH_REPO_NAME"));
+
+    // Act
+    const result = await act.runJob(jobId, {steps: {mock: [{name: "npm publish --access public", run: "echo 'publish step mocked'"}]}});
```
Comment (Ginxo) [inline test/.github/workflows/publish.test.ts:20]: this is the new mock feature

File: src/shared/util/fs-util.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    let fileContent: string;
+    try {
+        fileContent = fs.readFileSync(filePath, encoding);
+    } catch (e) {
+        throw new Error(`Error loading file from ${filePath}`)
+    }
+    try {
+        return YAML.parse(fileContent);
+    } catch (e) {
+        throw new Error(`Error parsing yaml from file ${filePath}`)
```
Comment (shubhbapna) [inline src/shared/util/fs-util.ts:21]: Same comment as `readFileSync`

File: test/.github/workflows/publish.test.ts (hunk 2 above)
Comment (shubhbapna) [reply test/.github/workflows/publish.test.ts:20]: Love it!

**Review by shubhbapna — CHANGES_REQUESTED** [review]
Some file structure changes to keep things consistent
- Remove `test/action-compiler/compiled.js`
- `src/shared/domain/github-workflow.types.ts` seems to be only used act `act` so I think it would make more sense to shift this file to `src/act`
- `src/shared/util/fs-util.ts` again seems to be used only by `act`. Should we shift it to act-helper? If not we can at least reduce the directory nesting and just have it as `src/util/fs-util.ts`?

Class consistency:
- For `src/act/act-helper.ts` why not have these functions as private methods of the `Act` class? If not should keep it consistent and convert this act-helper file from using functions to a class to keep things consistent?

File: src/act/act.ts (hunk 1 above)
Comment (Ginxo) [reply src/act/act.ts:87]: I agree, I will try that

**Comment by shubhbapna** [comment]
Moving `Act` to `act-js`. So won't be able merge this PR.

I will pick the non Act changes and open a separate PR

**Comment by shubhbapna** [comment]
fixed by #50

**Pull request closed without merging**